### PR TITLE
BIT gender difference scores 

### LINKED ISSUE #12: BIT refactor - calculate word frequency
After text cleaning (#6), BITs methodology includes
- [ ] Frequency of words in geography vs. computer science booklets was computed as:
word frequency = frequency of a word in all booklets / frequency of the specific part of speech type
in all booklets
- [ ] To extract words that are more frequent in Geography than Computer science booklets (and vice
versa), we subtracted word frequency in Computer science booklets from Geography booklets, and
identify top 20 with the highest difference.

### LINKED ISSUE #13: BIT refactor - word embeddings
After text cleaning (#6)  and finding word frequency (#12), BITs methodology includes:
- [x] Word embeddings: To assess masculinity/femininity of a word we used GloVe word embedding
method (Global Vectors for Word Representation), which maps all words into a vector space. We
used a pre-trained language model, which was trained in a general internet crawl on 6 billion texts.
- [x] We then computed the difference in average cosine similarity (using text2vec package in R)
between the target word, and the words {man, he, his, masculine, male} and {woman, she, her,
feminine, female}.

This could be written in such a way to easily substitute in different word embedding methods (which we plan to do as a further step)

### PR DESCRIPTION
**Pull request by lizgzil**
Closes #12 and #13 

This adds:
- Functionality to get word frequencies and the difference in these in the different corpii (`get_word_comparisons`)
- A class to calculate the gender score for a list of words (`GloveDistances`).
- Tests for these
- Updated readme and requirements
- Fixes a bit in the spell check to account for times there is no spelling suggestion

## Questions when we see their code
- This uses the pretrained glove model 6B.100d. I'm not sure which dimension BIT used, but we can change this accordingly if needed. This may account for why our numbers differ.
- At what point do they use the lemmas? I assume we wouldnt be able to find them in glove vocab, so not actually sure what they were found for?
- What do they do with words not in one corpus? I've given them a frequency of 0.
- Words can have multiple POS tags - I've chosen that the output POS tag for a word is the one most frequently given

## Usage with the CS and Geography corpuses

Tying this altogether (something like this - but neater!- will be put into a flow)
```
import json
import pandas as pd

from comp_sci_gender_bias.pipeline.process_text_utils import (
    TokenTagger,
    TextCleaner,
    GloveDistances,
    get_word_freq,
    get_word_comparisons
)

with open("inputs/data/compsci_descr.json", "r") as f:
    compsci_descr = json.load(f)
    
with open("inputs/data/geo_descr.json", "r") as f:
    geo_descr = json.load(f)

text_cleaner = TextCleaner()
token_tagger = TokenTagger()

comp_clean_text = [text_cleaner.clean(text) for text in compsci_descr.values()]
comp_clean_tagged = [token_tagger.tag(text) for text in comp_clean_text]
geo_clean_text = [text_cleaner.clean(text) for text in geo_descr.values()]
geo_clean_tagged = [token_tagger.tag(text) for text in geo_clean_text]

comp_clean_tagged_flatten = [c for cc in comp_clean_tagged for c in cc]
geo_clean_tagged_flatten = [c for cc in geo_clean_tagged for c in cc]

comp_df = pd.DataFrame({
    'Word': [c[0].lower() for c in comp_clean_tagged_flatten],
    'POS': [c[2] for c in comp_clean_tagged_flatten],
    'Corpus': ['CS']*len(comp_clean_tagged_flatten)
})
geo_df = pd.DataFrame({
    'Word': [c[0].lower() for c in geo_clean_tagged_flatten],
    'POS': [c[2] for c in geo_clean_tagged_flatten],
    'Corpus': ['Geo']*len(geo_clean_tagged_flatten)
})

word_differences_df = get_word_comparisons(comp_df, geo_df)

glove_dists = GloveDistances(glove_d=50)
glove_dists.load_glove2word2vec()

glove_sims = glove_dists.gender_similarity_difference_word_list(word_differences_df.index.tolist())

word_differences_df['Male - Female'] = word_differences_df.index.map(glove_sims)
```
Gives

||Geo - CS freq|	POS	|Word freq	|Word count| Male - Female |
|---|---|---|---|----|----|
|actively|	-0.000479	|ADV	|0.000039	|5|0.023567|
|anything	|0.001149	|PRON	|0.000047	|6|0.023038|
|joy	|0.000038	|NOUN	|0.000008|	1	|-0.049917|

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [ ] I have requested a code review

### CODE REVIEW

**Comment by lizgzil** [comment]
Our scores (100d model):
refactor_masc_scores

Our scores (50d model):
refactor_masc_scores_50


BITS:
Screenshot 2022-05-12 at 11 45 54

**Comment by lizgzil** [comment]
## POS = ['ADJ', 'ADV']
More common in geography:
`['geographical', 'physical', 'human', 'global', 'why', 'urban', 'natural', 'economic', 'well', 'tropical', 'coastal', 'also', 'newly', 'most', 'very', 'sustainable', 'local', 'highly', 'www', 'so']`

BIT found: `[geographic, physical, human, global, natural, urban, economic, living, coast, tropic]`

More common in CS:
`['fundamental', 'wired', 'logical', 'wireless', 'above', 'analytically', 'cultural', 'real', 'ultrahigh', 'able', 'robust', 'technical', 'logically', 'legal', 'creative', 'ethical', 'practical', 'digital', 'how', 'computational']`

BIT found: `[computational, how, digital, software, practical, ethical, legal, fundamental, robust, creative]`

## POS = ['NOUN']
More common in geography:
`['geography', 'world', 'fieldwork', 'environment', 'uk', 'challenges', 'issues', 'management', 'landscapes', 'section', 'hazards', 'change', 'climate', 'people', 'challenge', 'resource', 'topics', 'paper', 'processes', 'places']`

BIT found: `[geography, world, environment, fieldwork, challenge, issue, change, uk, study, landscape]`

More common in CS:
`['computers', 'solving', 'problem', 'security', 'course', 'problems', 'programs', 'software', 'future', 'technology', 'glossary', 'algorithms', 'study', 'computing', 'systems', 'work', 'science', 'programming', 'computer', 'bbc']`

BIT found: `[computer, science, problem, system, programming, program, programme, technology, course, algorithms]`

## POS = ['VERB']
More common in geography:
`['living', 'changing', 'do', 'include', 'live', 'making', 'including', 'based', 'studying', 'studied', 'surveying', 'take', 'affect', 'helps', 'extended', 'interpret', 'includes', 'investigating', 'collect', 'see']`

BIT found: `[include, live, make, face, explore, travel, collect, affect, examine, help]`

More common in CS:
`['working', 'programmed', 'given', 'goes', 'give', 'used', 'learn', 'set', 'works', 'designing', 'producing', 'write', 'create', 'using', 'develop', 'apply', 'solve', 'use', 'thinking', 'org']`

BIT found: `[solve, thinking, use, computing, apply, create, produce, give, develop, compute]`

File: comp_sci_gender_bias/tests/test_process_text.py [hunk 1]
```diff
@@ -30,3 +38,48 @@ def test_CleanText():
...
+
+    mother_score = glove_dists.gender_similarity_difference_word_list(["mother"])
+    father_score = glove_dists.gender_similarity_difference_word_list(["father"])
+
+    # At the very least 'mother' should be less masculine than 'father'
+    assert mother_score["mother"] < father_score["father"]
+
+    not_word_scores = glove_dists.gender_similarity_difference_word_list("notawordkd")
+
+    assert not_word_scores == None
```
Comment (Jack-0-0) [inline comp_sci_gender_bias/tests/test_process_text.py]:
```suggestion
    assert not_word_scores is None
```

**Review by Jack-0-0 — APPROVED** [review]
Thank you for this @lizgzil, the code runs as expected and the tests pass. It looks good to merge 👍

**Pull request merged**